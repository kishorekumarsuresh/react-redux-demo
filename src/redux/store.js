import { legacy_createStore as createStore ,applyMiddleware} from "redux";
import rootReducer from "./rootReducer";
//import logger from "redux-logger";
import { composeWithDevTools } from 'redux-devtools-extension';
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import thunk from "redux-thunk";

const persistConfig = {
  key : 'cake-count',
  storage,
  blacklist: ['ice'],
  whitelist : ['cake']
}     
const persistedReducer = persistReducer(persistConfig,rootReducer)
const store = createStore(persistedReducer,composeWithDevTools(applyMiddleware(thunk)))
const persistor = persistStore(store)
export default store
export {persistor}
