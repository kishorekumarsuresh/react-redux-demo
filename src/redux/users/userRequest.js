import axios  from "axios";

export function userRequest() {
    return axios.get("https://gorest.co.in/p/v2/users", {
      headers: {
        Authorizaton:
          "Bearer 0d1c0f5711b07c68f9aa845ecc00f75fc02f18e030c9d7e6040903b4968791e6",
      },
    })
    .then(response => {
      return Promise.resolve(response);
    })
    .catch(err => {
      return Promise.reject(err);
    })
}