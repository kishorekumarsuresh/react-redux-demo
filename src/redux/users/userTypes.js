export const SUCCESS_CALL = "SUCCESS_CALL"
export const FAILURE_CALL = "FAILURE_CALL"
