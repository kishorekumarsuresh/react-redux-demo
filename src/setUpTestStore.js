import { render } from "@testing-library/react"
import store, { persistor } from "./redux/store";
import { Provider } from "react-redux";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store,
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}


