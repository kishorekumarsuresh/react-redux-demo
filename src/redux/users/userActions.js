import { SUCCESS_CALL,FAILURE_CALL,CREATE_USER } from "./userTypes";
import axios from "axios";
import { userRequest } from "./userRequest";

export const successCall = (users) => {
  return{
    type:SUCCESS_CALL,
    payload : users,
  }
}

export const failureCall = (error) => {
  return{
    type:FAILURE_CALL,
    payload : error,
  }
}

export const createUser = (values) => {
  return{
    type:CREATE_USER,
    payload:values
  }
}

export const displayUsers = () => {
   return async (dispatch) => {
    // const dispatch = useDispatch()
    return userRequest().then(response =>{
      // return Promise.resolve(data)
      const values = response.data
      console.log('api suceeded', response.data)
       dispatch(successCall(values))
    })
    .catch(err => {
      const errMsg = err.message
      console.log("api fails",err)
      dispatch(failureCall(errMsg))
    })

  }
}

// export const newRegister = (obj) => {
//   const {name,gender,email} = obj
//   return (dispatch) =>{
//     dispatch(createUser(obj))
//    axios({
//     method:'POST',
//     url:'https://gorest.co.in/public/v2/users',
//     data: {name:name,gender:gender,email:email,status:'active'},
//     headers:{
//         'Authorization':"Bearer 0d1c0f5711b07c68f9aa845ecc00f75fc02f18e030c9d7e6040903b4968791e6"
//     }
//    })
//    .then(response=>{
//     console.log('success')
//     dispatch(successCall(response.data))
//    })
//    .catch(err => {
//     console.log('error')
//     dispatch(failureCall(err.message))
//    })
//   }
// }

