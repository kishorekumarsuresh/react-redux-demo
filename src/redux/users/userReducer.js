import { SUCCESS_CALL,FAILURE_CALL,CREATE_USER} from "./userTypes"

const initialState = {
  users : [],
  error : '',
  values : []
}

const userReducer = (state=initialState,action) => {
  switch(action.type){
    case SUCCESS_CALL : return {
      ...state,
      users : action.payload,
      error : '',
    }
    case FAILURE_CALL : return {
      ...state,
      users : [],
      error : action.payload,
    }
    case CREATE_USER: return{
      ...state,
      values : action.payload
  
    }
    default : return state
  }
}
export default userReducer
