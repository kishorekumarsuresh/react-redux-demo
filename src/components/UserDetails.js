import React,{useEffect} from 'react'
import {connect} from 'react-redux'
import { displayUsers } from '../redux/users/userActions'

function UserDetails({userData, displayUsers}) {
  useEffect(()=> {
    displayUsers()
    //dispatch(displayUsers())
  }, [])
  return (
    <div>
      <span data-testid="quote">UserDetails</span>
      {console.log('output', userData)}
      { userData.users ?
        userData.users.map((elem,index)=> <h2 key={index}>
         <p>{index+1}. {elem.name ? elem.name : 'Kishorekumar'}</p> 
          </h2>) :

        <div>
        No Data Found 
        </div>
    

      }
      <br></br>
      error - {userData.error}
    </div>
  )
}

const mapStateToProps = state => {
  return {
    userData : state.user,
    // err : state.error
  }
}
const mapDispatchToProps = dispatch => {
  return{
    displayUsers : () => dispatch(displayUsers())
  }
}
export default connect(mapStateToProps,mapDispatchToProps) (UserDetails)
