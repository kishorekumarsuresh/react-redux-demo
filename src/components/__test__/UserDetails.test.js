import { render, waitFor ,getByTestId, screen} from "@testing-library/react";
import UserDetails from "../UserDetails";
import { renderWithProviders } from 'setUpTestStore';
// import { displayUsers } from "../../redux/users/userActions";
// import { renderWithProviders } from "App.test";
//import * as api from '../redux/users/userActions'
import store, { persistor } from "../../redux/store";
import { userRequest } from "redux/users/userRequest";

jest.mock("../../redux/users/userRequest.js");

describe("Userdetails Component" , () => {
  beforeEach(() => {
    jest.clearAllMocks();
    // userRequest.mockImplementation(()=> Promise.resolve({
    //   data : [{name :"Jenni2"}],
    // }))
    
  });

  it ("checking api success call " , async () => {
    const{getByText, container} = renderWithProviders(<UserDetails/>, {store})
    await waitFor(()=>{
      expect(getByText("1. Jenni2")).toBeInTheDocument();
    });
    expect(container.firstChild).toMatchSnapshot();
  })

  
  it("checking usedetails in screen" , () => {
  
    renderWithProviders(<UserDetails/>, {store})
    const txt = screen.getByTestId("quote")
    expect(txt).toBeInTheDocument()
    
  })


  it.only("checking err msg in screen" ,async() => {
    userRequest.mockImplementation(()=> Promise.reject({
      error : {message :"error - Network Error"},
    }));
    renderWithProviders(<UserDetails/>, {store});
    await waitFor(()=>{
      expect("error - Network Error").toBeInTheDocument();
    })
  })

  // it ("checking api failure call " , async () => {
  //   userRequest.mockRejectedValue({})
  //   const{getByText} = renderWithProviders(<UserDetails/>, {store})
  //   await waitFor(()=>{
  //     expect(getByText("error")).toBeInTheDocument()
  //   })
  // })
})



