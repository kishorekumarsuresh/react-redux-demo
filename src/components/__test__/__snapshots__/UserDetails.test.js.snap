// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Userdetails Component checking api success call  1`] = `
<div>
  <span
    data-testid="quote"
  >
    UserDetails
  </span>
  <h2>
    <p>
      1
      . 
      Jenni2
    </p>
  </h2>
  <br />
  error - 
</div>
`;

exports[`Userdetails Component checking err msg in screen 1`] = `
<div>
  <span
    data-testid="quote"
  >
    UserDetails
  </span>
  <br />
  error - 
</div>
`;
